Query cache: Enable compression of cache entries

rschu1ze opened the pull request:
This PR changes two things in the query cache:

- ClickHouse reads table data in blocks of `max_block_size` rows but due to filtering, aggregation, etc., result blocks are typically much smaller than `max_block_size` (and sometimes bigger). The new setting `query_cache_squash_partial_results` (enabled by default) now controls if result blocks are squashed (if they are tiny) or split (if they are large) into blocks of `max_block_size` size before insertion into the query result cache. This reduces performance of writes into the query cache but improves compressability of cache entries and provides a "natural" block granularity when query results are later served from the query cache.

- Entries in the query cache are now compressed by default. This reduces the overall memory consumption at the cost of slower writes into / reads from the query cache. To disable compression, use the new setting `query_cache_compress_entries`.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Entries in the query cache are now squashed to max_block_size and compressed

alexey-milovidov commented on src/Interpreters/Cache/QueryCache.cpp:
```diff
@@ -218,6 +220,25 @@ void QueryCache::Writer::finalizeWrite()
 
     std::lock_guard lock(mutex);
 
+    auto to_single_chunk = [](const std::shared_ptr<Chunks>& chunks_) -> std::shared_ptr<Chunks>
```
If the resultset is large, making it a single chunk can be worse. Let's just squash to max_block_size.

rschu1ze replied:
Yepp, will check.

alexey-milovidov commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -1279,7 +1279,10 @@ The following settings are available:
 -   `max_entry_size`: The maximum size in bytes `SELECT` query results may have to be saved in the cache. Default value: `1048576` (1 MiB).
--   `max_entry_records`: The maximum number of records `SELECT` query results may have to be saved in the cache. Default value: `30000000` (30 mil).
+-   `max_entry_rows`: The maximum number of rows `SELECT` query results may have to be saved in the cache. Default value: `30000000` (30 mil).
+-   `compress_entries`: If cache entries are compressed. Reduces the cache memory consumption but reduces cache insert/read performance. Default value: `true`.
```
Another idea is to have the compressed flag near each entry, and then the compression can be controlled by a query-level setting on the side of the query writing to the cache.

rschu1ze replied:
Sounds good, can do.

rschu1ze committed:
Query cache: Enable compression and squashing of result blocks

ClickHouse reads table data in blocks of 'max_block_size' rows. Due to
filtering, aggregation, etc., result blocks are typically much smaller
than 'max_block_size' but there are also cases where they are much
bigger. Setting 'query_cache_squash_partial_results' (enabled by
default) now controls if result blocks are squashed (if they are tiny)
or split (if they are large) into blocks of 'max_block_size' size before
insertion into the query result cache. This reduces performance of
writes into the query cache but improves compressability of cache
entries and provides more natural block granularity when query results
are later served from the query cache.

Entries in the query cache are now also compressed by default. This
reduces the overall memory consumption at the cost of slower writes into
/ reads from the query cache. To disable compression, use setting
'query_cache_compress_entries'.

rschu1ze committed:
Merge branch 'master' into qc-compression

rschu1ze committed:
Merge branch 'master' into qc-compression

SmitaRKulkarni commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -726,7 +726,8 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                         std::make_optional<String>(context->getUserName()),
-                        std::chrono::system_clock::now() + std::chrono::seconds(settings.query_cache_ttl));
+                        /*dummy value for expires_at*/ std::chrono::system_clock::from_time_t(1),
+                        /*dummy value for is_compressed*/ true);
```
maybe we can set these as default values for function parameters

SmitaRKulkarni commented on src/Storages/System/StorageSystemQueryCache.cpp:
```diff
@@ -16,6 +16,7 @@ NamesAndTypesList StorageSystemQueryCache::getNamesAndTypes()
         {"key_hash", std::make_shared<DataTypeUInt64>()},
         {"expires_at", std::make_shared<DataTypeDateTime>()},
         {"stale", std::make_shared<DataTypeUInt8>()},
+        {"compressed", std::make_shared<DataTypeUInt8>()},
```
just curious, why we don't add this column to the end ? readability is affected after result size ?

rschu1ze replied:
Didn't think too much about the order :) But you are right, perhaps the important stuff (query, result_size, stale), should come first. I'd adjust that in a separate PR.

SmitaRKulkarni reviewed (commented):
Rest all LGTM

rschu1ze commented on tests/queries/0_stateless/02494_query_cache_squash_partial_results.sql:
```diff
@@ -0,0 +1,48 @@
+-- Run again to check that no bad things happen and that the result is as expected
+SELECT * FROM t ORDER BY c SETTINGS max_block_size = 2, use_query_cache = true;
+
+DROP TABLE t;
```
Yepp, that makes sense. Allow me to do that in a separate PR.

rschu1ze replied:
Ah, better let me push a follow-up commit into this PR.

rschu1ze committed:
Make the column order in system.query_cache more intutitive

rschu1ze committed:
Extend tests for compression of query cache entries

rschu1ze committed:
Merge remote-tracking branch 'origin/master' into qc-compression

CurtizJ commented:
See https://github.com/ClickHouse/ClickHouse/pull/48358#issuecomment-1494710268.

devcrafter commented on tests/queries/0_stateless/02494_query_cache_compression.sql:
```diff
@@ -0,0 +1,444 @@
+
+-- Create test table with lot's of rows
+CREATE TABLE t(c String) ENGINE=MergeTree ORDER BY c;
+INSERT INTO t values ('abc') ('def') ('abc') ('jkl');
```
Please
```sql
SELECT multiIf(n = 0, 'abc', n = 1, 'def', n = 2, 'abc', 'jkl')
FROM
(
    SELECT number % 4 AS n
    FROM numbers(4 * 400)
)
```

rschu1ze replied:
#48373, thanks.
